### PR TITLE
Convert makefiles to build LLVM/compiler-rt with CMake

### PR DESCRIPTION
**Pull request by brson**
This is certainly buggy, but I have successfully built on x86_64-unknown-linux-gnu and x86_64-pc-windows-gnu. I haven't built successfully on mac yet, and I've seen mysterious test failures on Linux, but I'm interested in throwing this at the bots to see what they think.

### CODE REVIEW

**Comment by brson** [comment]
FWIW re the mac failures: I was trying to build this on rust-mac6 and couldn't get it to recognize C++11 atomics. I'm not sure why this would be but I'm sort of crossing my fingers that clang is out of date there?

File: mk/rt.mk [hunk 1]
```diff
@@ -233,35 +233,60 @@ COMPRT_DEPS := $(wildcard \
...
+endif
+
+else
+COMPRT_LIB_NAME_$(1) := "$$(call CFG_STATIC_LIB_NAME_$(1),clang_rt.builtins-$$(COMPRT_ARCH_$(1)))"
+endif
+endif
+
+endif
+endif
+endif
```
Comment (alexcrichton) [inline mk/rt.mk]: This is totally a readable makefile now!

**Comment by shepmaster** [comment]
I tangentially asked this on the Internals forum, but why put the effort into updating the Makefiles instead of charging ahead with rustbuild?

**Comment by brson** [comment]
@shepmaster @alexcrichton says we're not going to be ready to remove the makefiles soon, and I want to have emscripten support soon.

**Comment by brson** [comment]
@alexcrichton why aren't we going to be able to remove the makefiles soon?

**Comment by brson** [comment]
Exporting MACOSX_DEPLOYMENT_TARGET=10.7 didn't help.

**Comment by brson** [comment]
I figured out that the `-stdlib=libc++` flag makes `#import <atomic>` work, but can't see how to make the cmake atomic config check pass that flag. `-DLLVM_ENABLE_LIBCXX=ON` doesn't do it.

**Comment by brson** [comment]
@alexcrichton does rustbuild definitely work on os x?

**Comment by brson** [comment]
Er, I can run cmake by hand successfully, so I must have the script messed up.

**Comment by brson** [comment]
OK, I think it must have been a corrupted local configuration. Seems to be working now.

**Comment by shepmaster** [comment]
I am definitely building using rustbuild on OS X.

**Comment by alexcrichton** [comment]
Some messages on IRC ---

```
15:52 <~acrichto> brson: yeah rustbuild osx works (gated)
15:52 <~acrichto> you're right about -stdlib though, that's the flag not 
                  DEPLOYMENT_VERSION
15:52 <~acrichto> and it all comes back...
15:52 <~acrichto> brson: if you mirror the rustbuild cmake exactly it should in 
                  theory work
15:53 <~acrichto> brson: I think it's this -- 
https://github.com/rust-lang/rust/blob/master/src/bootstrap/build/mod.rs#L808
15:53 <~acrichto> you may have already found that...
```

But yeah I was misremembering MACOSX_DEPLOYMENT_VERSION, it's just -stdlib in CFLAGS somehow

**Comment by brson** [comment]
I've built successfully on OS X and am going to send this through the bots to see what happens.

@bors r+

**Comment by sanxiyn** [comment]
It seems that this will break Travis, because Ubuntu package installs LLVMConfig.cmake in a place different from what compiler-rt build system expects. I am trying to figure out how to unbreak.

**Comment by alexcrichton** [comment]
@brson I [hate shell](http://stackoverflow.com/questions/411495/build-argument-lists-containing-whitespace)

**Comment by eddyb** [comment]
[Actual failure, on android](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/9462/steps/configure/logs/stdio).

**Comment by alexcrichton** [comment]
guh

**Comment by alexcrichton** [comment]
An update on this:

* Fixed a few bugs related to `--enable-ccache`, both in the `./configure` script and when building compiler-rt
* Fixed a bunch of things when building Android libraries
* Updated compiler-rt to ensure we can build an aarch64 iOS library, this should be off the `release_38` branch like LLVM is
* Ensure we pass `-Wno-error` to compiler-rt to not translate warnings to errors
* Fixed 32-bit OSX

Still working on lots more hiccups

**Comment by alexcrichton** [comment]
So I've been working through tons of failures in dev, and oh my god I swear that the incantation to build compiler-rt is different on every single platform. There is now an absurd amount of logic in mk/rt.mk just to figure out how to build compiler-rt.

And it's not even done!

**Comment by alexcrichton** [comment]
I think this is finally ready to go. If I could quantify the amount of money it took to test this in dev I'd probably be sad, so I won't do that!

Need to restart buildbot to pick up some changes to the images, then I'll r+

**Comment by alexcrichton** [comment]
guh and of course travis is broken...

**Comment by alexcrichton** [comment]
Ok, some more hacks that were applied:

* In the docker image on Travis we now somewhat-randomly symlink some folders to coerce compiler-rt to pick up the system LLVM correctly.
* A few patches were applied to compiler-rt to coerce it to compile:
  * working on powerpc - https://github.com/rust-lang/compiler-rt/commit/a1ef94b76029780a510bc2dc9c6a791bd091ff19
  * working on msvc - https://github.com/rust-lang/compiler-rt/commit/dd79ba2f6c9f06c3aa3f3e33bc6f8450fe433322

One glaring hole is that `make -j10` does **not** compile LLVM in parallel right now. I have no idea why and I'm still trying to figure that out...

**Comment by alexcrichton** [comment]
Ok, parallelism should be fixed

**Comment by alexcrichton** [comment]
@bors: retry

* weird lldb error on OSX?

**Comment by alexcrichton** [comment]
error may be related to https://github.com/rust-lang/rust/issues/33391

**Comment by alexcrichton** [comment]
@bors: retry

* umm... Contention?

**Comment by alexcrichton** [comment]
@bors: retry

Looks like it's stuck compiling stage1 libsyntax but I can't seem to reproduce it locally. Hoping that it's just contention locally...

**Comment by brson** [comment]
Wow I'm shocked this got through. Thanks for fixing it @alexcrichton

**Pull request merged**